### PR TITLE
[DOC-REVIEW] Frames README (#2)

### PR DESCRIPTION
**Pull request by Sharon-iguazio**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,221 +4,501 @@
...
-3. Stream
-4. CSV - for testing purposes
+<a id="client-python-api-reference"></a>
+## Client Python API Reference
 
+- [Overview](#overview)
+- [User Authentication](#user-authentication)
+- [Client Constructor](#client-constructor)
+- [Common Client Method Parameters](#client-common-method-params)
+- [The create Method](#method-create)
```
Comment (talIguaz) [inline README.md]: I Would just call it `Create`, `Write`, `Read`, etc... (instead of "the create method")

File: README.md [hunk 2]
```diff
@@ -4,221 +4,501 @@
...
+```
+Then, you need to create and initialize an instance of the `Client` class; see [Client Constructor](#client-constructor).
+You can then use the client methods to perform different data operations on the supported backend types:
+
+<a id="client-methods"></a>
+#### Client Methods
+
+The `Client` class features the following methods for supporting basic data operations:
+
+- [`create`](#method-create) &mdash; create a new NoSQL or TSDB table or a stream ("the backend").
```
Comment (talIguaz) [inline README.md]: Create for NoSQL is not supported

File: README.md [hunk 3]
```diff
@@ -4,221 +4,501 @@
...
+- [`delete`](#method-delete) &mdash; delete the backend.
```
Comment (talIguaz) [inline README.md]: "Delete the backend" sounds harsh.. the previous line is "create a new *NoSQL* or *TSDB table* or a *stream*", but this deletes the "backend". maybe rephrase this.
Maybe - "Delete a backend's table"

File: README.md [hunk 4]
```diff
@@ -4,221 +4,501 @@
...
+- [`read`](#method-read) &mdash; read data from the backend (as a pandas DataFrame or DataFrame iterator).
+- [`write`](#method-write) &mdash; write one or more DataFrames to the backend.
+- [`execute`](#method-execute) &mdash; execute a command on the backend. Each backend may support multiple commands.
```
Comment (talIguaz) [inline README.md]: "Execute a backend specific command. Each backend may support multiple commands."

File: README.md [hunk 5]
```diff
@@ -4,221 +4,501 @@
...
 
-#### backend specific parameters
-##### TSDB
-* rate 
-* aggregates (optional)
-* aggregation-granularity (optional)
+<a id="method-create-params-tsdb"></a>
+#### tsdb Backend create Parameters
+
+- <a id="method-create-tsdb-param-rate"></a>**rate** (Required) &mdash; `string` &mdash; the ingestion rate TSDB's metric-samples, as `"[0-9]+/[smh]"` (where `s` = seconds, `m` = minutes, and `h` = hours); for example, `"1/s"` (one sample per minute).
```
Comment (talIguaz) [inline README.md]: ... the approximate ingestion rate for the TSDB's metric-samples ...

File: README.md [hunk 6]
```diff
@@ -4,221 +4,501 @@
...
+
+  - **Type:** String
+  - **Requirement:** Required
+
+Additional method-specific parameters are described for each method.
+
+<a id="method-create"></a>
+### create Method
+
+Creates a new data collection (table/stream) in a platform data container according to the configured backend.
```
Comment (talIguaz) [inline README.md]: According to the *given* backend.
Configured sounds static and complicated, in frames you specify the backend on every function call. 
(same for all other methods)

File: README.md [hunk 7]
```diff
@@ -4,221 +4,501 @@
...
     Requirement IG-5339, & DOC IG-12272. The original write example below had
     an `expression` parameter:
-v3c.write(backend='kv', table='mytable', dfs=df, expression='city="NY"', condition='age>14')
+v3c.write(backend="kv", table="mytable", dfs=df, expression="city='NY'", condition="age>14")
     -->
-* condition=' ' (optional) - for detailed information on condition expressions see [docs](https://www.iguazio.com/docs/reference/latest-release/expressions/condition-expression/)
+  <!-- [IntInfo] [c-no-update-expression-support] (24.9.19) See Bug IG-12510,
+    Requirement IG-5339, & DOC IG-12272. -->
+
+- <a id="method-write-kv-param-condition"></a>**condition** (Optional) (default: none) &mdash; a platform condition expression that defines conditions for performing the update.
```
Comment (talIguaz) [inline README.md]: Also we support `Expression` (iv'e noticed we didnt document this before, not sure if it was on purpose).
Basicaaly the `expression` is just a regular update expression.

File: README.md [hunk 8]
```diff
@@ -4,221 +4,501 @@
...
-* max_in_message: int - Maximal number of rows per message
-* sharding_keys: []string (Experimental)- list of specific sharding keys to query. For range scan formatted tables only.
-* segments: []int64 (Not yet supported)
-* total_segments: int64 (Not yet supported)
-* sort_key_range_start: string (Not yet supported)
-* sort_key_range_end: string (Not yet supported)
+<a id="method-read-backend-common-params"></a>
+#### Common read Parameters
+
+All Frames backends that support the `read` method support the following common parameters, which can be set in the `attrs` method parameter:
```
Comment (talIguaz) [inline README.md]: `which can be set in the `attrs` method parameter:` -> that's just not true. remove this.

File: README.md [hunk 9]
```diff
@@ -4,221 +4,501 @@
...
-#### backend specific parameters
-##### TSDB
-* start: string - delete since start
-* end: string - delete since start
+<a id="method-delete"></a>
+### delete Method
+
+Deletes a data collection (table/stream) in a platform data container according to the configured backend.
+<br/>
+The `kb` backend also supports an optional [`filter`](#method-delete-kv-param-filter) parameter that can be used to delete only specific items in a NoSQL tables.
```
Comment (talIguaz) [inline README.md]: `kb` -> `kv`

File: README.md [hunk 10]
```diff
@@ -4,221 +4,501 @@
...
+- [`stream` backend `read` parameters](#method-read-params-stream)
+
+Reads data from a backend.
+
+<a id="method-read-syntax"></a>
+#### Syntax
+
 ```python
-df = client.read(backend='tsdb', query="select avg(cpu) as cpu, avg(diskio), avg(network)from mytable", start='now-1d', end='now', step='2h')
+read(backend=<backend>, table=<table>, attrs=<backend-specific parameters>)
```
Comment (talIguaz) [inline README.md]: `attrs=<backend-specific parameters>` -> that's not really the syntax. The backends specific parameters are just written as normal parameters. See the examples you wrote below for reference.

File: README.md (hunk 1 above)
Comment (Sharon-iguazio) [reply README.md]: @talIguaz  I'd rather leave it as-is, because this is reference documentation and the current format clarifies the data type documented in each section + it's consistent with our current doc-site reference doc convetions. (Note that the method names are written with lowercase letters because these are their actual names in the code.)

File: README.md (hunk 2 above)
Comment (Sharon-iguazio) [reply README.md]: @talIguaz Good catch. I forgot to update the method descriptions in the overview. => I'll prepare an update to the descriptions of all the methods in the overview.

File: README.md (hunk 3 above)
Comment (Sharon-iguazio) [reply README.md]: You're right, "delete backend" is misleading. It was based on the original doc in the tutorials Frames NB. => I've now edited all the method descriptions both in the overview and in the method reference sections, including to bypass this. Also, per your input, I replaced all "configured backend" references with "specified backend type". I'll commit the changes shortly to the PR branch.

File: README.md (hunk 5 above)
Comment (Sharon-iguazio) [reply README.md]: @talIguaz the next line in this parameter description is "The rate should be calculated according to the slowest expected ingestion rate.". Doesn't this cover what you wanted to add?

File: README.md (hunk 6 above)
Comment (Sharon-iguazio) [reply README.md]: I edited the method description, including to refer to "the specified backend type". The changes will be committed soon to the PR branch.

File: README.md (hunk 4 above)
Comment (talIguaz) [reply README.md]: Dont think "custom" fits here. since the user can not create his own custom commands as the word "custom" implies.

File: README.md (hunk 7 above)
Comment (Sharon-iguazio) [reply README.md]: We did document `expression` and as part of my recent changes I commented-out this documentation because Golan and Adi decided that we're not ready to support update expressions. See the internal `[c-no-update-expression-support]` comments that I added in the README when I filtered out the update-expression doc.

File: README.md (hunk 8 above)
Comment (Sharon-iguazio) [reply README.md]: Removed; will be committed shortly to the PR branch.

File: README.md (hunk 9 above)
Comment (Sharon-iguazio) [reply README.md]: Fixed; will be committed shortly to the PR branch.

File: README.md (hunk 10 above)
Comment (Sharon-iguazio) [reply README.md]: I fixed the syntax of all methods parameters; will be committed shortly to the PR branch + I added documentation of the common `write` `attrs` parameter and `execute` `args` parameter; will be committed shortly to the PR branch.

File: README.md (hunk 4 above)
Comment (Sharon-iguazio) [reply README.md]: OK, I changed "custom" to backend-specific, as you suggested, both in the README and in the help text in the code.

**Comment by Sharon-iguazio** [comment]
@talIguaz I finished the second review round and committed the changes to the PR. In addition to changes pertaining to your review comments, I also made additional fixes and improvements — including to add return-value sections for the client constructor and the read method, change the syntax code snippets, and fix and improve the method parameter descriptions (to be continued ...).
I also edited the help text in the client code files (again, only the first round).

File: clients/py/v3io_frames/client.py [hunk 11]
```diff
@@ -25,18 +25,22 @@
...
         Parameters
         ----------
         address : str
-            framesd server address
+            Address of the Frames service (framesd)
         session : Session
             Session object
         frame_factory : class
-            DataFrame factory (currencly pandas and cudf supported)
+            DataFrame factory; currently, pandas and cuDF are supported
```
Comment (talIguaz) [inline clients/py/v3io_frames/client.py]: Not sure we want to say that cuDF is officially supported yet

File: README.md (hunk 5 above)
Comment (talIguaz) [reply README.md]: Don't think `slowest` is necessarily better. i guess an average rate will suite better.
But it depends on if there's a big difference between the expected rates. if so we need to think what will be better, otherwise i guess avg should work.

File: README.md [hunk 12]
```diff
@@ -4,221 +4,627 @@
...
+    limit=0, data_format='', row_layout=False, max_in_message=0, marker='',
+    iterator=False, **kw)
+```
+
+<a id="method-read-common-params"></a>
+#### Common read Parameters
+
+All Frames backends that support the `read` method support the following common parameters:
+
+- <a id="method-read-param-iterator"></a>**iterator** &mdash; (Optional) (default: `False`) &mdash; `bool` &mdash; `True` to return a DataFrames iterator; `False` to return a single DataFrame.
```
Comment (talIguaz) [inline README.md]: better to order the parameters with the `optional` ones in the end

File: README.md [hunk 13]
```diff
@@ -4,221 +4,627 @@
...
+ test backend. -->
+
+<a id="method-create-common-params"></a>
+#### Common create Parameters
+
+All Frames backends that support the `create` method support the following common parameters:
+
+- <a id="method-create-param-attrs"></a>**attrs** &mdash; A dictionary of `<argument name>: <value>` pairs for passing additional backend-specific parameters (arguments).
+
+  - **Type:** dict
```
Comment (talIguaz) [inline README.md]: The style of the parameter description here is different than the other methods (read,write..). is it on purpose?

File: README.md (hunk 5 above)
Comment (Sharon-iguazio) [reply README.md]: @talIguaz currently, in our doc-site TSDB CLI tutorial, Frames v3io/tutorials NB, and Frames README file we say that the TSDB ingestion rate should be set according to the slowest expected ingestion rate. This doc was added based on Golan's guidelines. @dinal please let me know whether we need to change to this guideline, in light of Tal's input, and if so, what do you want the new guideline to be.

File: clients/py/v3io_frames/client.py (hunk 11 above)
Comment (Sharon-iguazio) [reply clients/py/v3io_frames/client.py]: I agree. @talIguaz @dinal feel free to remove the cuDF reference from the help text; note that it's not included in the README. (As I didn't write the original help text, I didn't want to take it upon myself to remove the cuDF reference from this text and I didn't have time to ask last week.)

File: README.md (hunk 12 above)
Comment (Sharon-iguazio) [reply README.md]: @talIguaz I plan to reorder all parameter lists alphabetically and include separate requirement bullets (as I did in the constructor documentation). I just didn't have time to handle this for this PR + I want to do the reordering in a separate commit to simplify diffs.

File: README.md (hunk 13 above)
Comment (Sharon-iguazio) [reply README.md]: @talIguaz ultimately, all parameter descriptions should have separate type and requirement bullets (and valid and default values bullets where relevant), I just didn't have time to do it for this PR.
